server/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -444,6 +450,20 @@ public long sizeOfGensAboveSeqNoInBytes(long minSeqNo) {
         }
     }
 
+    private long earliestGenerationLastModifiedAge(long minGeneration) {
+        try (ReleasableLock ignored = readLock.acquire()) {
+            ensureOpen();
+            Optional<? extends BaseTranslogReader> earliestGeneration = Stream.concat(readers.stream(), Stream.of(current))
+                .filter(r -> r.getGeneration() >= minGeneration)
+                .min(Comparator.comparingLong(BaseTranslogReader::getGeneration));
```

Comment:
can we maybe map this to a stream of longs via getLastModifiedTime and collect using `Collectors.minBy(Comparator.comparing(BaseTranslogReader:: getLastModifiedTime))` or something similar? also, can we also take the writer into account for completeness?

Reply:
To be clear - I think we want to know how the oldest file is, regardless of the generations. It will always be the oldest generation and the first in the reader list, but I don't think we want to rely on it. Part of the role of the stats is to validate things are correct.

Reply:
Sure but we can't use `BaseTranslogReader:: getLastModifiedTime` as the method throws a checked exception.

Does `Stream.concat(readers.stream(), Stream.of(current))` not include the writer?

Reply:
> Sure but we can't use BaseTranslogReader:: getLastModifiedTime as the method throws a checked exception.

Fair enough. No streams for us - we need to do it the old fashion way :D 

> Does Stream.concat(readers.stream(), Stream.of(current)) not include the writer?

Yes. Current is a TranslogWriter.

Reply:
I've settled on the following WDYT?
```java
    long earliestLastModifiedAge() {
        try (ReleasableLock ignored = readLock.acquire()) {
            ensureOpen();
            return findEarliestLastModifiedAge(System.currentTimeMillis(), Stream.concat(readers.stream(), Stream.of(current)));
        } catch (ElasticsearchException e) {
            throw new TranslogException(shardId, e.getDetailedMessage(), e);
        }
    }

    /**
     * Returns the age of the oldest entry in the translog files in seconds
     */
    static long findEarliestLastModifiedAge(long currentTime, Stream<BaseTranslogReader> baseTranslogReaders) {
        return currentTime - baseTranslogReaders.mapToLong(Translog::uncheckedGetLastModifiedTime).min().orElse(currentTime);
    }

    private static long uncheckedGetLastModifiedTime(BaseTranslogReader r) {
        try {
            return r.getLastModifiedTime();
        } catch (IOException e) {
            throw new ElasticsearchException("Unable to get the earliest last modified time for the transaction log");
        }
    }
```

On the testing side I've added:
```java
    public void testFindEarliestLastModifiedAge() throws IOException {
        long fixedTime = System.currentTimeMillis();
        long earlyTime = fixedTime - 10000;
        BaseTranslogReader early = mock(BaseTranslogReader.class);
        stub(early.getLastModifiedTime()).toReturn(earlyTime);
        long earlierTime = fixedTime - 20000;
        BaseTranslogReader earlier = mock(BaseTranslogReader.class);
        stub(earlier.getLastModifiedTime()).toReturn(earlierTime);
        long earliestTime = fixedTime - 30000;
        BaseTranslogReader earliest = mock(BaseTranslogReader.class);
        stub(earliest.getLastModifiedTime()).toReturn(earliestTime);
        Stream<BaseTranslogReader> readers = Stream.of(early, earlier, earliest);
        assertThat(Translog.findEarliestLastModifiedAge(fixedTime, readers), equalTo(earliestTime));
    }
```

./gradlew check running now, feel free to comment soonest :D

Reply:
I would rather not have the uncheckedGetLastModifiedTime method. Instead just do old school iteration on the reader list, keeping the min of the last modified time. Then do a one off extra min with the writer's value.

Re the test - I would prefer if you make an array of random positive long values, map readers to them and make sure the method returns the right value when compared to a calculation that is made directly on the array.